### DISCUSSION TITLE
How to use *GLSL* vertex and fragment shaders with `pygfx`?

### DISCUSSION THREAD

**Discussion by KelSolaar** [pygfx/pygfx#598] [Q&A]
Hello,

At first glance it does not seem trivial to use *GLSL* vertex and fragment shaders with `pygfx`.

`wgpu-py` has some example on how to do that, e.g. https://github.com/pygfx/wgpu-py/blob/main/examples/triangle_glsl.py, but I would be keen to use a higher level interface possibly so that it is possible to benefit from all the good work that you guys did!
Cursory look at some of the `pygfx` shader code, seems to imply that only *wgsl* is supported.

For context, I'm trying to port some of our GL code, i.e. https://github.com/AcademySoftwareFoundation/OpenColorIO/blob/main/src/apps/ocioview/ocioview/viewer/image_plane.py, to `pygfx` but we do need to use *GLSL* as it is what *OpenColorIO* emits.

Some hints at the best path forward would be appreciated!

Cheers,

Thomas

**Comment by almarklein** [comment]
At the moment we indeed expect shaders for world objects to be written in WGSL. Apart from that we also impose a few restrictions, e.g. fragment shaders must return a `FragmentOutput` struct.  Plus there's some other auto-generated shader code, e.g. for uniforms buffers and storage buffers).

So if we were to support GLSL shaders, we'd have to duplicate all that for GLSL, which I something I hope we can avoid ...

Some thoughts:
* Use [Naga](https://github.com/gfx-rs/naga) to convert your GLSL to WGSL. And then somehow massage that code so it fits in the shape that pygfx expects.
* We could introduce an option to ditch the boilerplate code, so the shader simply provides the whole shader as-is. This would break some stuff, in particular the blending (transparency) system, but for custom viz this might be worth it. This might be the easiest way forward for you, but I don't foresee all consequences yet.
* In the end, if the shader produces GLSL instead of WGSL wgpu-py should pick this up and compile it correctly.

**Reply by KelSolaar** [reply]
Thanks Almar!

I considered transpiling the code but it looks like a good rabbit hole. The second point is looking more like what I was thinking about whilst reading the code: e.g. modifying things here https://github.com/pygfx/pygfx/blob/main/pygfx/renderers/wgpu/_pipeline.py#L79 so that GLSL input is possible.

Cheers,

Thomas

**Reply by almarklein** [reply]
Taking another look, I think having point two is actually easier than I thought! The shader classes all implement a `get_code()` method that combines all the code pieces together. They usually include `self.code_definitions()` and `self.code_common()`, but they strictly speaking don't *have* to. And if it only return GLSL it *should* just work!

**Reply by KelSolaar** [reply]
Correct me if I'm wrong but the shader code ultimately goes through the [generate_wgsl](https://github.com/pygfx/pygfx/blob/4b15625b1fdee9b17117b5fdfeafc22cbf1b2f1b/pygfx/renderers/wgpu/_shaderbase.py#L334) definition which in turns call the `resolve_varyings` and `resolve_depth_output` definitions that do quite a lot of text massaging that is potentially incompatible with GLSL?

**Reply by almarklein** [reply]
Aah yes, you're correct, I overlooked these two lines. So we'd need to somehow skip these steps.
